src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs

```diff
@@ -86,7 +86,15 @@ public Library GetLibrary(LibraryRange libraryRange, NuGetFramework targetFramew
             // This must exist in the external references
             if (_externalProjectsByUniqueName.TryGetValue(name, out ExternalProjectReference externalReference))
             {
-                packageSpec = externalReference.PackageSpec;
+                if (externalReference.PackageSpec == null ||
+                    libraryRange.VersionRange.FindBestMatch(new NuGetVersion[] { externalReference.PackageSpec.Version }) != null)
+                {
+                    packageSpec = externalReference.PackageSpec;
+                }
+                else
+                {
+                    externalReference = null;
+                }
```

Comment:
I reverted this change in my local copy of your PR branch and changed the line at https://github.com/NuGet/NuGet.Client/blob/6a096d0504851db695bef483801a992d65262c64/src/NuGet.Core/NuGet.DependencyResolver.Core/PackagingUtility.cs#L40 to the following:

```cs
TypeConstraint = LibraryDependencyTarget.Package
```

I'm not sure why we set TypeConstraint = LibraryDependencyTarget.PackageProjectExternal in the method that is trying to identify the dependency from the nuspec. I would assume that the dependencies downloaded from the nuspec are always of the 'package' type. I tried changing that line of code as I mentioned above, and the test passed. It's possible that I am missing some context here.

Reply:
My assumption is that this is done to make the "Project over Package" feature work.

Reply:
Looking at the blame report, this change was introduced few years ago. https://github.com/NuGet/NuGet.Client/commit/c7b2acdfd0f2594e3e796c5d3cfa86af62909794

Reply:
Yep, from the commit title, this very much sounds like the "Project over Package" feature.

Reply:
@ViktorHofer is correct, it's for the project over package feature.

Reply:
Unresolved this conversation as I have a follow up question. If it is related to project over package feature, does it answer Andy's question in https://github.com/NuGet/NuGet.Client/pull/5397#pullrequestreview-1613151646 especially `But if I understand correctly, say I have Contoso.Utilities as a transitive package, if I clone the repo and add Contoso.Utilities.csproj to my solution, then the build will use that project's dll? I can make changes to Contoso.Utilities and test with my app, without neeing to publish packages or changing dependency versions?
If so, that's very cool, and brings us closer to what we had with project.json back in the dnx/.NET Core 1.0 preview days.`

Reply:
That scenario already works, but now it accounts for the versions.
The build will use the project only if the version matches.

Reply:
Thanks Nikolche.